Transmute big endian `s6_addr` and `[u16; 8]`

tesuji opened the pull request:
The old code already made the assumption to reinterpret
`Ipv6Addr` as `[u16; 8]`.

Glibc, Linux, FreeBSD, Win32 all makes this assumption.
The main motivation of using union it to better optimize code.
Godbolt: https://rust.godbolt.org/z/sdP4Pc
Const is introducing unsafe when transmuting.

ref:
* https://docs.microsoft.com/en-us/windows/win32/api/in6addr/ns-in6addr-in6_addr
* https://github.com/freebsd/freebsd/blob/1d6e4247415d264485ee94b59fdbc12e0c566fd0/contrib/ntp/lib/isc/include/isc/ipv6.h#L63
* https://github.com/zephyrproject-rtos/zephyr/blob/8b531aa996bba254c03129658490af59597acd78/include/net/net_ip.h#L137
* https://sourceware.org/git/?p=glibc.git;a=blob;f=inet/netinet/in.h;h=f6355c7efe5192b88337b136ef687fe9a5ed648c;hb=HEAD#l216

pickfire commented on library/std/src/net/ip.rs:
```diff
@@ -1042,6 +1042,20 @@ impl From<[u8; 4]> for IpAddr {
+// This union assumes that all fields are interpreted as big endian.
+// Also, all of its field has the same size.
+#[repr(C)]
+union in6_union {
```
Do we need the name `_union`?

lzutao replied:
I took the same name in C sources. Now I've changed it to Ipv6Bytes which may be more idiomatic.

lzutao commented:
Wow. Lots of tests are falling. I should reconsider this approach.

lzutao commented on library/std/src/net/ip.rs:
```diff
@@ -1056,27 +1070,12 @@ impl Ipv6Addr {
-                    h as u8,
-                ],
+                // SAFETY: Each element of `s6_addr16` is big edian.
+                s6_addr: unsafe { in6_union { s6_addr16: [a, b, c, d, e, f, g, h] }.s6_addr },
```
The assumption here is wrong about arguments: (`a`, `b`, ...).
Those are not guaranteed to be Big Endian.

cuviper replied:
Right, the parameters are native-endian, and the output from `segments` is native too.

Maybe you could use `[a.to_be_bytes(), b.to_be_bytes, ...]` to get `[[u8; 2]; 8]` instead, and then cast that into `s6_addr`. But once we're dealing with byte swaps, your godbolt advantage may disappear, so I think this will only be a matter of code style.

retep998 replied:
Or `[a.to_be(), b.to_be(), ...]` to get a `[u16; 8]` directly that is big endian.

cuviper replied:
This actually does lead to better instructions: https://rust.godbolt.org/z/dv4Pje

retep998 commented:
> All members of the IN6_ADDR structure must be specified in network-byte-order (big-endian).

I'm pretty sure this means that if you're on a little endian system you have to byte swap your shorts before writing them to this structure.

lzutao commented:
Thanks both. I will resolve it later.
Do you have an interest to take a look at #75110 ?
In that PR, I use newtype to enforce endianness for Ipv4Addr. But beware that it is a draft one
and I'm figuring out how a lot of tests are falling.

retep998 commented on library/std/src/net/ip.rs:
```diff
@@ -1119,17 +1128,8 @@ impl Ipv6Addr {
-            u16::from_be_bytes([arr[14], arr[15]]),
-        ]
+        // SAFETY: All element of `s6_addr16` must be big edian.
+        unsafe { Ipv6Bytes { addr: self.inner.s6_addr }.addr16 }
```
This will need to do something like `[u16::from_be(arr[0]), u16::from_be(arr[1]), ...]`

lzutao replied:
Thank you. Realized this mistake an hour ago. Will fix it.

pickfire commented on library/std/src/net/ip.rs:
```diff
@@ -1056,27 +1070,22 @@ impl Ipv6Addr {
-                    (h >> 8) as u8,
-                    h as u8,
-                ],
+                // SAFETY: All element of `addr16` are big edian.
```
```suggestion
                // SAFETY: All element of `addr16` are big endian.
```

pickfire commented on library/std/src/net/ip.rs:
```diff
@@ -1042,6 +1042,20 @@ impl From<[u8; 4]> for IpAddr {
+//
+// # Safety
+//
+// This union assumes that all fields are interpreted as big endian.
+// Also, all of its field has the same size.
```
Should we add a test for this?

pickfire commented on library/std/src/net/ip.rs:
```diff
@@ -1042,6 +1042,20 @@ impl From<[u8; 4]> for IpAddr {
+// This union assumes that all fields are interpreted as big endian.
+// Also, all of its field has the same size.
+#[repr(C)]
+union Ipv6Bytes {
```
I don't quite understand why do we need to declare this union ourselves? Why not just use the bytes directory rather than going through a layer of union like `[a.to_be(), b.to_be(), ...]`?

lzutao replied:
I don't quite get your point. What is bytes directory? Is the old code (in master)  using bytes directory approach?

pickfire replied:
Oh, I mean as in "use the bytes directly", sorry typo.

lzutao replied:
Ah, thanks for explaining. Here the codegen diff: https://rust.godbolt.org/z/cE9zqd

pickfire replied:
Oh, the binary looks much shorter. Looks like performance is the only reason, should we explain that union is used for performance reasons here?

lzutao replied:
Good point. Will do.

tesuji committed:
Transmute between big endian `s6_addr` and `[u16; 8]`.

The old code already made the assumption to reinterpret
`Ipv6Addr` as `[u16; 8]`.

Glibc, Linux, FreeBSD, Win32 all makes this assumption.
The main motivation of using union it to better optimize code.

ref:
* https://docs.microsoft.com/en-us/windows/win32/api/in6addr/ns-in6addr-in6_addr
* https://github.com/freebsd/freebsd/blob/1d6e4247415d264485ee94b59fdbc12e0c566fd0/contrib/ntp/lib/isc/include/isc/ipv6.h#L63
* https://github.com/zephyrproject-rtos/zephyr/blob/8b531aa996bba254c03129658490af59597acd78/include/net/net_ip.h#L137
* https://sourceware.org/git/?p=glibc.git;a=blob;f=inet/netinet/in.h;h=f6355c7efe5192b88337b136ef687fe9a5ed648c;hb=HEAD#l216

Co-authored-by: Josh Stone <cuviper@gmail.com>
Co-authored-by: Peter Atashian <retep998@gmail.com>

pickfire commented on library/std/src/net/ip.rs:
```diff
@@ -1884,10 +1896,16 @@ impl From<[u16; 8]> for IpAddr {
 
+    #[test]
+    fn ipv6bytes_all_fields_have_the_same_size() {
+        assert_eq!(size_of::<[u8; 16]>(), size_of::<[u16; 8]>());
```
Shouldn't this be linked to the fields like `Ipv6Bytes::addr` and `Ipv6Bytes::addr16`? It may be possible that users set the union to some other types and this test case remains the same as it is currently but it was actually broken.

lzutao replied:
Eh, it cannot be because addr and addr16 are not associated types. 
But I do have a way to check fields size, even though it contains unsafe.

pickfire replied:
I think even using unsafe is better than leaving this hanging, having no relationship with the original union.

lzutao replied:
Oops, I found that I don't even need unsafe.

pickfire commented on library/std/src/net/ip.rs:
```diff
@@ -1884,10 +1899,19 @@ impl From<[u16; 8]> for IpAddr {
+        use super::Ipv6Bytes;
+        let _addr = Ipv6Bytes { addr: [0_u8; 16] };
+        let _addr16 = Ipv6Bytes { addr16: [0_u16; 8] };
+        assert_eq!(size_of::<Ipv6Bytes>(), size_of::<[u8; 16]>());
```
I wonder if `size_of_val` works here? Like `size_of_val(&addr)` and `size_of_val(&addr16)`.

lzutao replied:
Does they have the same type?

pickfire replied:
Yes, they have the same type. If they have the same type it means they have the same size? I am not sure the behavior of `size_of_val` for a `union`. Looks like we could do that with unsafe. Let me send a suggestion.

Prototype: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=8c5f14837cefa7cc61951bd338278eec

pickfire replied:
```suggestion
        unsafe {
            assert_eq!(size_of_val(&addr.addr), size_of_val(&addr16.addr16));
        }
```

lzutao replied:
> If they have the same type it means they have the same size? 
Yes, if not that's would be the most unsound issue.

The two `let` lines above will fail at complie time if the field changes its size.

pickfire replied:
> Yes, if not that's would be the most unsound issue.

That is what we want to check in the test here rather than typing it out as text in comment since the type system does not seemed to be able to validate this.

> The two let lines above will fail at complie time if the field changes its size.

Yes, that is what we want to check, both the fields should have the same size.

cuviper replied:
Comparing `size_of_val` on two instances of the union will always be equal, but comparing on each field makes sense.

pickfire replied:
> Comparing size_of_val on two instances of the union will always be equal, but comparing on each field makes sense.

@cuviper Yes, I did a test in nightly to check, comparing the same type give the same output but however when I compare the fields, it gives different output. We could use that here, I rather it does not need unsafe though but seemed like we need it.

pickfire commented on library/std/src/net/ip.rs:
```diff
@@ -1884,10 +1899,19 @@ impl From<[u16; 8]> for IpAddr {
 
+    #[test]
+    fn ipv6bytes_all_fields_have_the_same_size() {
+        use super::Ipv6Bytes;
+        let _addr = Ipv6Bytes { addr: [0_u8; 16] };
+        let _addr16 = Ipv6Bytes { addr16: [0_u16; 8] };
+        assert_eq!(size_of::<Ipv6Bytes>(), size_of::<[u8; 16]>());
```
```suggestion
        use super::Ipv6Bytes;
        let addr = Ipv6Bytes { addr: [0_u8; 16] };
        let addr16 = Ipv6Bytes { addr16: [0_u16; 8] };
        unsafe {
            assert_eq!(size_of_val(&addr.addr), size_of_val(&addr16.addr16));
        }
```

pickfire commented on library/std/src/net/ip.rs:
```diff
@@ -1884,10 +1899,19 @@ impl From<[u16; 8]> for IpAddr {
 // Tests for this module
 #[cfg(all(test, not(target_os = "emscripten")))]
 mod tests {
+    use crate::mem::size_of;
```
```suggestion
    use crate::mem::{size_of, size_of_val};
```

cuviper commented on library/std/src/net/ip.rs:
```diff
@@ -1056,27 +1074,22 @@ impl Ipv6Addr {
-                    (h >> 8) as u8,
-                    h as u8,
-                ],
+                // SAFETY: All elements in `addr16` are big endian.
```
Being big-endian isn't really relevant to *safety*, just correctness of the API. For memory safety in the `unsafe` union access, what matters is that the fields are all the same size and have the same validity, so initializing one field is always valid data for the other field.

cuviper commented:
Instead of the temporary union, have you tried `transmute`? (with `#![feature(const_fn_transmute)]`)
It seems to codegen identically, so LLVM even deduplicates these: https://rust.godbolt.org/z/1Khbee

lzutao commented:
> transmute already includes the size assertion.

Where is it defined? I can only find this:
https://github.com/rust-lang/rust/blob/f50f1c8e17a34ccaa0263c637e9686492b79477f/src/librustc_codegen_ssa/mir/block.rs#L1343-L1380

lzutao commented:
> Instead of the temporary union, have you tried `transmute`?

It would work. But we may lose all safety comments that documented in Ipv6Bytes.

pickfire commented:
> It would work. But we may lose all safety comments that documented in Ipv6Bytes.

We could move those safety comments to unsafe, the good thing is it will be shorter, the bad thing is hard to test and track the unsafe (probably).

lzutao commented:
> We could move those safety comments to unsafe transmute

Could be. However there are 2 "transmute/unsafe" site, I don't like duplicating it 
as it would be harder/easier to forget to synchronize in the future.

cuviper commented:
Transmute size is checked here:
https://github.com/rust-lang/rust/blob/8bc801b05019cd3e0ef19e6c4c028d55baa645d2/src/librustc_passes/intrinsicck.rs#L71-L124

> But we may lose all safety comments that documented in Ipv6Bytes.

I don't understand what you're afraid to lose. The big-endian safety point is already repeated in the two functions, and `transmute` enforces size itself, better than a comment.

lzutao commented:
Nice, I was afraid of using transmute because its documentation states that "transmute is incredibly unsafe".

cuviper commented:
I suspect everything that makes transmute dangerous is also a hazard for union type punning. At least in this case we're only talking about benign types without niches/padding/etc.

lzutao commented:
Test passed locally. Changed to using transmute.

pickfire reviewed (approved):
Looks good to me.

cuviper commented on library/std/src/net/ip.rs:
```diff
@@ -1045,27 +1046,23 @@ impl Ipv6Addr {
-                    h as u8,
-                ],
+                // All elements in `addr16` are big endian.
+                // SAFETY: `[u16; 8]` is always safe to transmuted to `[u8; 16]`.
```
```suggestion
                // SAFETY: `[u16; 8]` is always safe to transmute to `[u8; 16]`.
```

cuviper commented on library/std/src/net/ip.rs:
```diff
@@ -1108,16 +1105,19 @@ impl Ipv6Addr {
     #[stable(feature = "rust1", since = "1.0.0")]
     pub fn segments(&self) -> [u16; 8] {
-        let arr = &self.inner.s6_addr;
+        // All elements in `s6_addr` must be big edian.
+        // SAFETY: `[u8; 16]` is always safe to transmuted to `[u16; 8]`.
```
```suggestion
        // All elements in `s6_addr` must be big endian.
        // SAFETY: `[u8; 16]` is always safe to transmute to `[u16; 8]`.
```

cuviper reviewed (changes requested):
Please update the PR title to "transmute".

cuviper commented:
@bors r+

bors commented:
:pushpin: Commit 0210fd3d7303a626f6e821bde8fdfea056cc043b has been approved by `cuviper`

<!-- @bors r=cuviper 0210fd3d7303a626f6e821bde8fdfea056cc043b -->
<!-- homu: {"type":"Approved","sha":"0210fd3d7303a626f6e821bde8fdfea056cc043b","approver":"cuviper"} -->

bors commented:
:hourglass: Testing commit 0210fd3d7303a626f6e821bde8fdfea056cc043b with merge 6b1e00131ee7682486cfc5f45ffb1966b89d4a81...
<!-- homu: {"type":"BuildStarted","head_sha":"0210fd3d7303a626f6e821bde8fdfea056cc043b","merge_sha":"6b1e00131ee7682486cfc5f45ffb1966b89d4a81"} -->

bors commented:
:broken_heart: Test failed - [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=35893)
<!-- homu: {"type":"BuildFailed","builder_url":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=35893","builder_name":"checks-azure"} -->

lzutao commented:
Hmm, macOS builds didn't start.

cuviper commented:
@bors retry rollup
